consensus/hotstuff/committees/cluster_committee.go

```diff
@@ -0,0 +1,115 @@
+)
+
+// Cluster represents the committee for a cluster of collection nodes. Cluster
+// committees are epoch-scoped.
+//
+// Clusters build blocks on a cluster chain but must obtain identity table
+// information from the main chain. Thus, block ID parameters in this Committee
+// implementation reference blocks on the cluster chain, which in turn reference
+// blocks on the main chain - this implementation manages that translation.
+type Cluster struct {
+	state    protocol.ReadOnlyState
+	payloads storage.ClusterPayloads
+	me       flow.Identifier
+	// pre-computed leader selection for the full lifecycle of the cluster
+	selection *leader.LeaderSelection
+	// a filter that returns all members of the cluster committee allowed to vote
+	clusterMemberFilter flow.IdentityFilter
+	// initial set of cluster members, WITHOUT updated weight
+	initialClusterMembers flow.IdentityList
+}
+
+func NewClusterCommittee(
+	state protocol.ReadOnlyState,
+	payloads storage.ClusterPayloads,
+	cluster protocol.Cluster,
+	epoch protocol.Epoch,
+	me flow.Identifier,
+) (*Cluster, error) {
+
+	selection, err := leader.SelectionForCluster(cluster, epoch)
+	if err != nil {
+		return nil, fmt.Errorf("could not compute leader selection for cluster: %w", err)
+	}
+
+	com := &Cluster{
+		state:                 state,
+		payloads:              payloads,
+		me:                    me,
+		selection:             selection,
+		clusterMemberFilter:   cluster.Members().Selector(),
+		initialClusterMembers: cluster.Members(),
+	}
+	return com, nil
+}
+
+func (c *Cluster) Identities(blockID flow.Identifier, selector flow.IdentityFilter) (flow.IdentityList, error) {
+
+	// first retrieve the cluster block payload
+	payload, err := c.payloads.ByBlockID(blockID)
+	if err != nil {
+		return nil, fmt.Errorf("could not get cluster payload: %w", err)
+	}
+
+	// if a reference block is specified, use that
+	if payload.ReferenceBlockID != flow.ZeroID {
+		identities, err := c.state.AtBlockID(payload.ReferenceBlockID).Identities(filter.And(
+			selector,
+			c.clusterMemberFilter,
+		))
+		return identities, err
+	}
+
+	// otherwise, this is a root block, in which case we use the initial cluster members
+	return c.initialClusterMembers.Filter(selector), nil
+}
+
+func (c *Cluster) Identity(blockID flow.Identifier, nodeID flow.Identifier) (*flow.Identity, error) {
+
+	// first retrieve the cluster block payload
+	payload, err := c.payloads.ByBlockID(blockID)
+	if err != nil {
+		return nil, fmt.Errorf("could not get cluster payload: %w", err)
+	}
+
+	// if a reference block is specified, use that
+	if payload.ReferenceBlockID != flow.ZeroID {
+		identity, err := c.state.AtBlockID(payload.ReferenceBlockID).Identity(nodeID)
+		if err != nil {
+			return nil, fmt.Errorf("could not get identity for node (id=%x): %w", nodeID, err)
+		}
+		if !c.clusterMemberFilter(identity) {
+			return nil, fmt.Errorf("node (id=%x) is not a valid cluster committee member", nodeID)
+		}
+	}
```

Comment:
do we need to return something in this `if` case?

Reply:
Yes, thank you. Fixed in https://github.com/onflow/flow-go/pull/157/commits/c10a1ad6f8d0359a14ccb4121f7184e7fe9d8734